lib/engine/game/g_steam_over_holland/game.rb

```diff
@@ -356,25 +361,40 @@ def non_floated_corporations
         end
 
         def check_distance(route, visits, train = nil)
-          super
-
           raise GameError, 'Route cannot begin/end in a town' if visits.first.town? || visits.last.town?
+
+          super
         end
 
         def revenue_for(route, stops)
-          super
+          revenue = super
 
-          raise GameError, 'Route visits same hex twice' if route.hexes.size != route.hexes.uniq.size
+          abilities = abilities(route.corporation, :hex_bonus)
+          return revenue if abilities.empty?
+
+          bonus_hexes = abilities.flat_map(&:hexes)
+          bonus = 20 * stops.map(&:hex).map(&:coordinates).intersect(bonus_hexes).size
+          revenue + bonus
         end
 
         def sell_shares_and_change_price(bundle, allow_president_change: true, swap: nil, movement: nil)
```

Comment:
This method could be tidied up.

All three branches of the `if` statement call `@stock_market.move_left` but only one branch uses the `price_drop` variable you set at the start of the method.

Could this be rewritten to have the `if` statement modify `price_drop` as required, moving the call to `move_left` out of the branches. And move the `@round.issued_shares[corporation] = true` to `GSteamOverHollad::Step::BuyTrain.process_sell_shares` to match where it's set in the `IssueShares` step.

Reply:
Alright, I cleaned up that section quite a bit, thanks for the suggestion. I also made a change to revenue_for, because I realized that with the suggested code, abilities is only an array if the corp has a hex_bonus ability. So now the gate on that is `return revenue unless abilities`. Tested, works.